evm/src/recursive_verifier.rs

```diff
@@ -1,148 +1,574 @@
+            .collect();
+        start += 4 * (1 << config.fri_config.cap_height);
+        let ctl_challenges = GrandProductChallengeSet {
+            challenges: (0..config.num_challenges)
+                .map(|i| GrandProductChallenge {
+                    beta: v[start + 2 * i],
+                    gamma: v[start + 2 * i + 1],
+                })
+                .collect(),
+        };
+        start += 2 * config.num_challenges;
+        let challenger_state_before = v[start..start + SPONGE_WIDTH].try_into().unwrap();
+        let challenger_state_after = v[start + SPONGE_WIDTH..start + 2 * SPONGE_WIDTH]
+            .try_into()
+            .unwrap();
+
+        start += 2 * SPONGE_WIDTH;
+        let ctl_zs_last = v[start..].to_vec();
+
+        Self {
+            trace_cap,
+            ctl_zs_last,
+            ctl_challenges,
+            challenger_state_before,
+            challenger_state_after,
+        }
+    }
+}
+
+impl<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>
+    RecursiveAllProof<F, C, D>
+{
+    /// Verify every recursive proof.
+    pub fn verify(
+        self,
+        verifier_data: &[VerifierCircuitData<F, C, D>; NUM_TABLES],
+        inner_config: &StarkConfig,
+    ) -> Result<()>
+    where
+        [(); C::Hasher::HASH_SIZE]:,
+    {
+        let pis: [_; NUM_TABLES] = std::array::from_fn(|i| {
+            PublicInputs::from_vec(&self.recursive_proofs[i].public_inputs, inner_config)
+        });
+
+        let mut challenger = Challenger::<F, C::Hasher>::new();
+        for pi in &pis {
+            for h in &pi.trace_cap {
+                challenger.observe_elements(h);
+            }
+        }
+        let ctl_challenges =
+            get_grand_product_challenge_set(&mut challenger, inner_config.num_challenges);
+        // Check that the correct CTL challenges are used in every proof.
+        for pi in &pis {
+            ensure!(ctl_challenges == pi.ctl_challenges);
+        }
+
+        challenger.duplexing();
+        let state = challenger.state();
+        ensure!(state == pis[0].challenger_state_before);
+        // Check that the challenger state is consistent between proofs.
+        for i in 1..NUM_TABLES {
+            ensure!(pis[i].challenger_state_before == pis[i - 1].challenger_state_after);
+        }
+
+        // Verify the CTL checks.
+        let degrees_bits = std::array::from_fn(|i| verifier_data[i].common.degree_bits);
+        verify_cross_table_lookups::<F, C, D>(
+            self.cross_table_lookups,
+            pis.map(|p| p.ctl_zs_last),
+            degrees_bits,
+            ctl_challenges,
+            inner_config,
+        )?;
+
+        // Verify the proofs.
+        for (proof, verifier_data) in self.recursive_proofs.into_iter().zip(verifier_data) {
+            verifier_data.verify(proof)?;
+        }
+        Ok(())
+    }
+
+    /// Recursively verify every recursive proof.
+    pub fn verify_circuit<W>(
```

Comment:
I think we should separate the circuit building and witness generation this is doing, so that e.g. a verify-only user can build the circuit without having to do any witness generation or proving.

Reply:
So right now I've designed this function to be a gadget, since it might be used as part of a larger circuit, so there's no proving involved. Do you think it makes sense to have a copy of this function without the witness generation?